server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -354,9 +349,10 @@ private static ShardGenerations buildGenerations(SnapshotsInProgress.Entry snaps
     }
 
     private static Metadata metadataForSnapshot(SnapshotsInProgress.Entry snapshot, Metadata metadata) {
+        final Metadata.Builder builder;
```

Comment:
This change is a little ugly because now we filter data streams that aren't part of the snapshot from metadata but not indices. In fact, for indices we could (and probably should) do the same as the index metadata for indices that aren't recorded in `SnapshotInfo` is not touched by restores anyway and just pointless bytes.
For data streams we absolutely have to clean up the metadata though since when writing `SnapshotInfo`, we use the metadata contents to determine what data streams go into `SnapshotInfo`.
If we don't clean this up here, then data streams not initially request for snapshotting will be part of the snapshot for every snapshot that includes global metadata.

Reply:
I don't see another way around this too. Although not a super clean change, but it is needed.